lib/pure/endians.nim

```diff
@@ -98,16 +123,438 @@ else:
+    copyMem(outp, inp, 4)
+
+  proc littleEndian16*(outp, inp: pointer){.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    copyMem(outp, inp, 2)
+
+  proc bigEndian64*(outp, inp: pointer) {.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    swapEndian64(outp, inp)
+
+  proc bigEndian32*(outp, inp: pointer) {.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    swapEndian32(outp, inp)
+
+  proc bigEndian16*(outp, inp: pointer) {.inline, deprecated:
+      "Deprecated since v1.1.0; use `fromBytesXX/toBytesXX` instead".} =
+    swapEndian16(outp, inp)
+
+
+# New API starts here
+
+when not useBuiltinSwap or defined(testing) or defined(nimdoc):
+  func slowSwap16(a: uint16): uint16 {.inline.} =
+    result = (a shl 8) or (a shr 8)
+
+  func slowSwap32(a: uint32): uint32 {.inline.} =
+    var a = ((a shl  8) and 0xff00ff00'u32) or
+            ((a shr  8) and 0x00ff00ff'u32)
+    result = (a shl 16) or (a shr 16)
+
+  func slowSwap64(a: uint64): uint64 {.inline.} =
+    var a = ((a shl 8) and 0xff00ff00ff00ff00'u64) or
+            ((a shr 8) and 0x00ff00ff00ff00ff'u64)
+    a = ((a shl 16) and 0xffff0000ffff0000'u64) or
+        ((a shr 16) and 0x0000ffff0000ffff'u64)
+    result = (a shl 32) or (a shr 32)
+
+when not useBuiltinSwap:
+  template swap16(a: uint16): uint16 = slowSwap16(a)
+  template swap32(a: uint32): uint32 = slowSwap32(a)
+  template swap64(a: uint64): uint64 = slowSwap64(a)
+
+type
+  SomeInteger8* = int8|uint8
+    ## Type class matching 8-bit integer types.
+
+  SomeInteger16* = int16|uint16
+    ## Type class matching 16-bit integer types.
+
+  SomeInteger32* = int32|uint32
+    ## Type class matching 32-bit integer types.
+
+  SomeInteger64* = int64|uint64
+    ## Type class matching 64-bit integer types.
+    ##
+  SomeNumber32* = SomeInteger32|float32
+    ## Type class matching all 32-bit number types.
+
+  SomeNumber64* = SomeInteger64|float64
+    ## Type class matching all 64-bit number types.
+
+
+func swapEndian*[T: SomeNumber](value: T): T {.inline, mySince: (1,1).} =
+  ## Swaps the byte order of a number.
+  runnableExamples:
+    doAssert swapEndian(0xdeadbeef'i32) == 0xefbeadde'i32
+
+  when T is SomeInteger8:  value
+  elif T is SomeInteger16: cast[T](swap16(cast[uint16](value)))
+  elif T is SomeNumber32:  cast[T](swap32(cast[uint32](value)))
+  elif T is SomeNumber64:  cast[T](swap64(cast[uint64](value)))
+
+
+proc swapEndian*(T: typedesc[SomeNumber],
+                 buf: pointer) {.inline, mySince: (1,1).} =
+  ## Swaps the byte order of a number at a memory location in-place. The type
+  ## of the number has to be passed in as an argument.
+  runnableExamples:
+    var i = 0xdeadbeef'i32
+    swapEndian(int32, i.addr)
+    doAssert i == 0xefbeadde'i32
+
+  var value = cast[ptr T](buf)[]
+  cast[ptr T](buf)[] = swapEndian(value)
```

Comment:
Cool, I'll make it work with NimVM then too somehow. I'm not using compile-time evaluation much myself, but I see that it can useful plus at least I'll learn a bit more about NimVM in the process. Back to work... 😄